core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java

```diff
@@ -0,0 +1,282 @@
+                                                Arrays.toString(
+                                                        bestCandidateReplicaIds.toArray())));
+
+                                // try to migrate one of best replica candidate
+                                for (ReplicaID bestCandidateReplicaId : bestCandidateReplicaIds) {
+                                    // TODO check if recently migrated (check for cool down
+                                    //  time)
+                                    if (migrate(bestCandidateReplicaId)) {
+                                        timer.reset();
+                                        return;
+                                    }
+                                }
+                            }
+                        },
+                        METRIC_WATCH_INTERVAL);
+    }
+
+    /** Starts metric watcher */
+    public void start() {
+        timer.start();
+    }
+
+    private boolean migrate(ReplicaID replicaID) {
+        InetSocketAddress currentKernelServer;
+        try {
+            // get replica instance event handler
+            EventHandler replica = microServiceManager.getReplicaDispatcher(replicaID);
+            currentKernelServer = replica.getHost();
+        } catch (Exception e) {
+            logger.warning("Failed to get replica dispatcher");
+            e.printStackTrace();
+            return false;
+        }
+
+        // predict new kernel server
+        InetSocketAddress kernelServer =
+                predictor.getBestKernelServer(replicaID, currentKernelServer);
+        logger.fine(
+                String.format(
+                        "Replica [%s] should migrate from %s to %s",
+                        replicaID, currentKernelServer, kernelServer));
+
+        // skipping migration if same kernel server selected
+        if (kernelServer.equals(currentKernelServer)) {
+            logger.info(
+                    "Skipping migration as predicted kernel server is same as replica kernel server");
+            return false;
+        }
+
+        // trigger migration
+        return triggerMigration(replicaID, kernelServer);
+    }
+
+    private synchronized List<ReplicaID> getBestCandidates(List<ReplicaID> replicas) {
+        List<ReplicaID> bestCandidates = new ArrayList<ReplicaID>();
+
+        // calculate processing time for each replica and add them in sorted
+        // heap
+        for (ReplicaID replicaID : replicas) {
+            try {
+                replicaHeap.add(
+                        new AbstractMap.SimpleEntry<ReplicaID, Long>(
+                                replicaID, aggregatedProcessingTime(replicaID)));
+            } catch (MicroServiceNotFoundException e) {
+                logger.warning(
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service not available",
+                                replicaID));
+                e.printStackTrace();
```

Comment:
Printing a stack trace to the console is generally not useful.  Rather log appropriately, or throw an exception (which is probably not appropriate in this case).